pkg/ingestor/parser/parser.go

```diff
@@ -0,0 +1,286 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/verifier"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/sigstore/sigstore/pkg/cryptoutils"
+)
+
+const (
+	algorithmSHA256 string = "sha256"
+)
+
+type docTreeBuilder struct {
+	identities    []assembler.IdentityNode
+	graphBuilders []*graphBuilder
+}
+
+type graphBuilder struct {
+	doc                   *processor.Document
+	foundIdentities       []assembler.IdentityNode
+	foundSubjectArtifacts []assembler.ArtifactNode
+	foundAttestations     []assembler.AttestationNode
+	foundBuilders         []assembler.BuilderNode
+	foundDependencies     []assembler.ArtifactNode
+}
+
+func newDocTreeBuilder() *docTreeBuilder {
+	return &docTreeBuilder{
+		identities:    []assembler.IdentityNode{},
+		graphBuilders: []*graphBuilder{},
+	}
+}
+
+func newGraphBuilder() *graphBuilder {
+	return &graphBuilder{
+		foundIdentities:       []assembler.IdentityNode{},
+		foundSubjectArtifacts: []assembler.ArtifactNode{},
+		foundAttestations:     []assembler.AttestationNode{},
+		foundBuilders:         []assembler.BuilderNode{},
+		foundDependencies:     []assembler.ArtifactNode{},
+	}
+}
+
+// ParseDocumentTree takes the DocumentTree and create graph inputs (nodes and edges) per document node
+func ParseDocumentTree(docTree processor.DocumentTree) ([]assembler.AssemblerInput, error) {
+
+	assemblerinputs := []assembler.AssemblerInput{}
+	docTreeBuilder := newDocTreeBuilder()
+	err := docTreeBuilder.parse(docTree)
+	if err != nil {
+		return []assembler.AssemblerInput{}, err
+	}
+	for _, builder := range docTreeBuilder.graphBuilders {
+		assemblerinput := builder.createAssemblerInput(docTreeBuilder.identities)
+		assemblerinputs = append(assemblerinputs, assemblerinput)
+	}
+
+	return assemblerinputs, nil
+}
+
+func (b *graphBuilder) createAssemblerInput(foundIdentities []assembler.IdentityNode) assembler.AssemblerInput {
+	assemblerinput := assembler.AssemblerInput{
+		Nodes: b.createNodes(),
+		Edges: b.createEdges(foundIdentities),
+	}
+	return assemblerinput
+}
+
+func (b *graphBuilder) createEdges(foundIdentities []assembler.IdentityNode) []assembler.GuacEdge {
+	edges := []assembler.GuacEdge{}
+	for _, i := range foundIdentities {
+		for _, a := range b.foundAttestations {
+			edges = append(edges, assembler.IdentityForEdge{IdentityNode: i, AttestationNode: a})
+		}
+	}
+	for _, s := range b.foundSubjectArtifacts {
+		for _, b := range b.foundBuilders {
```

Comment:
nit: use different variable so that it's not confusing to have two versions of `b`